File: account-service/src/main/java/com/koriebruh/accountservice/event/TransferEvent.java
package com.koriebruh.accountservice.event;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;

/**
 * Kafka event payload consumed from topic: transfer.event
 * Published by transaction-service after a transfer is processed.
 *
 * reference_id is the idempotency key — used to prevent double-posting
 * in balance_ledger on Kafka redelivery.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TransferEvent {

    @JsonProperty("reference_id")
    private String referenceId;         // idempotency key

    @JsonProperty("source_account_number")
    private String sourceAccountNumber; // account to debit (negative amount)

    @JsonProperty("target_account_number")
    private String targetAccountNumber; // account to credit (positive amount)

    @JsonProperty("amount")
    private BigDecimal amount;          // always positive, direction handled by sign

    @JsonProperty("currency")
    private String currency;

    @JsonProperty("occurred_at")
    private Instant occurredAt;
}

File: transfer-service/internal/event/transfer_event.go
package event

import (
	"time"

	"github.com/shopspring/decimal"
)

// TransferEvent is published to Kafka topic: transfer.event
// Consumed by account-service to update balances.
type TransferEvent struct {
	UserCode            string          `json:"user_code"`             // for bushiness logic like send email
	ReferenceID         string          `json:"reference_id"`          // idempotency key
	SourceAccountNumber string          `json:"source_account_number"` // account to debit
	TargetAccountNumber string          `json:"target_account_number"` // account to credit
	Amount              decimal.Decimal `json:"amount"`                // always positive
	Currency            string          `json:"currency"`              // source currency
	OccurredAt          time.Time       `json:"occurred_at"`
}
